Index: src\states\pointnclick.c
===================================================================
--- src\states\pointnclick.c
+++ src\states\pointnclick.c
@@ -10,8 +10,9 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "meta_tiles.h"
 
 #ifndef POINT_N_CLICK_CAMERA_DEADZONE
 #define POINT_N_CLICK_CAMERA_DEADZONE 24
@@ -25,8 +26,9 @@
     camera_deadzone_x = POINT_N_CLICK_CAMERA_DEADZONE;
     camera_deadzone_y = POINT_N_CLICK_CAMERA_DEADZONE;
     PLAYER.dir = DIR_RIGHT;
     actor_set_anim(&PLAYER, ANIM_CURSOR);
+    dynamic_actor_init();
 }
 
 void pointnclick_update(void) BANKED {
     UBYTE angle, hit_trigger, is_hover_trigger, is_hover_actor;
@@ -110,5 +112,6 @@
             // Run trigger script
             trigger_interact(hit_trigger);
         }
     }
+    dynamic_actor_update();
 }
