Index: src\core\actor.c
===================================================================
--- src\core\actor.c
+++ src\core\actor.c
@@ -16,9 +16,10 @@
 #include "collision.h"
 #include "ui.h"
 #include "vm.h"
 #include "macro.h"
-#include "continuous_scene.h"
+#include "data/states_defines.h"
+#include "scene_transition.h"
 
 #ifdef STRICT
     #include <gb/bgb_emu.h>
     #include <gb/crash_handler.h>
@@ -71,8 +72,9 @@
     player_moving           = FALSE;
     player_iframes          = 0;
     player_collision_actor  = NULL;
     emote_actor             = NULL;
+
     memset(actors, 0, sizeof(actors));
 }
 
 void player_init(void) BANKED {
@@ -205,12 +207,20 @@
         }
 
         if (CHK_FLAG(actor->flags, ACTOR_FLAG_PINNED)) {
             screen_x = SUBPX_TO_PX(actor->pos.x);
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+            screen_y = SUBPX_TO_PX(actor->pos.y - actor->pos_z);
+#else
             screen_y = SUBPX_TO_PX(actor->pos.y);
+#endif
         } else {
             screen_x = SUBPX_TO_PX(actor->pos.x) - draw_scroll_x;
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+            screen_y = SUBPX_TO_PX(actor->pos.y - actor->pos_z) - draw_scroll_y;
+#else
             screen_y = SUBPX_TO_PX(actor->pos.y) - draw_scroll_y;
+#endif
         }
 
         if (((window_hide_actors) && (((screen_x + 8) > WX_REG) && ((screen_y - 8) > WY_REG)))) {
             continue;
@@ -549,5 +559,6 @@
     } else if (player_iframes != 0) {
         player_iframes--;
     }
     player_collision_actor = NULL;
+    check_transition_to_scene_collision();
 }
\ No newline at end of file
