Index: include\collision.h
===================================================================
--- include\collision.h
+++ include\collision.h
@@ -4,24 +4,55 @@
 #include <gbdk/platform.h>
 
 #include "math.h"
 #include "bankdata.h"
+#include "data/states_defines.h"
 
 #define COLLISION_TOP 0x1
 #define COLLISION_BOTTOM 0x2
 #define COLLISION_LEFT 0x4
 #define COLLISION_RIGHT 0x8
 #define COLLISION_ALL 0xF
 #define TILE_PROP_LADDER 0x10
 
+#define METATILE_SIZE_8     0
+#define METATILE_SIZE_16    1
+
+#if METATILE_SIZE == METATILE_SIZE_16
+
+#define METATILE_X_OFFSET(x) (x >> 1)
+#define METATILE_Y_OFFSET(y) ((y >> 1) << image_tile_width_bit)
+#define TILE_X_OFFSET(x) (x & 1)
+#define TILE_Y_OFFSET(y) ((y & 1) << 5)
+#define METATILE_MAP_OFFSET(x, y)  (METATILE_Y_OFFSET(y) + METATILE_X_OFFSET(x))
+#define TILE_MAP_OFFSET(metatile_idx,x,y)  (UWORD)(((metatile_idx & 0xF0) << 2) + ((metatile_idx & 15) << 1) + TILE_Y_OFFSET(y) + TILE_X_OFFSET(x))
+#define MAX_MAP_DATA_SIZE 0x1C00
+#define COLLISION_DATA_SIZE 0x0400
+
+#else
+
+#define METATILE_Y_OFFSET(y) (y << image_tile_width_bit)
+#define METATILE_MAP_OFFSET(x, y)  (METATILE_Y_OFFSET(y) + x)
+#define MAX_MAP_DATA_SIZE 0x1F00
+#define COLLISION_DATA_SIZE 0x0100
+
+#endif
+
+#define SRAM_MAP_DATA_PTR 0xA000
+#define SRAM_COLLISION_DATA_PTR (SRAM_MAP_DATA_PTR + MAX_MAP_DATA_SIZE)
+
 extern UBYTE collision_bank;
 extern unsigned char *collision_ptr;
 extern UBYTE image_tile_width;
 extern UBYTE image_tile_height;
 
 extern UBYTE tile_hit_x;
 extern UBYTE tile_hit_y;
 
+extern uint8_t __at(SRAM_COLLISION_DATA_PTR) sram_collision_data[COLLISION_DATA_SIZE];
+extern uint8_t __at(SRAM_MAP_DATA_PTR) sram_map_data[MAX_MAP_DATA_SIZE];
+extern UBYTE metatile_collision_bank;
+extern UBYTE image_tile_width_bit;
 extern UBYTE scene_transition_enabled;
 
 /**
  * Check if point is within positioned bounding box.
@@ -63,10 +94,19 @@
  * @param ty Top tile
  * @return Tile value, 0 if no collisions, COLLISION_ALL if out of bounds
  */
 inline UBYTE tile_at(UBYTE tx, UBYTE ty) {
-    if ((tx < image_tile_width) && (ty < image_tile_height))
-        return ReadBankedUBYTE(collision_ptr + (ty * (UINT16)image_tile_width) + tx, collision_bank);
+    if ((tx < image_tile_width) && (ty < image_tile_height)) {
+        if (metatile_collision_bank) {
+#if METATILE_SIZE == METATILE_SIZE_16
+            return sram_collision_data[TILE_MAP_OFFSET(sram_map_data[METATILE_MAP_OFFSET(tx, ty)], tx, ty)];
+#else
+            return sram_collision_data[sram_map_data[METATILE_MAP_OFFSET(tx, ty)]];
+#endif
+        } else {
+            return ReadBankedUBYTE(collision_ptr + (ty * (UINT16)image_tile_width) + tx, collision_bank);
+        }
+    }
     return (scene_transition_enabled)? 0: COLLISION_ALL;
 }
 
 /**
@@ -78,9 +118,9 @@
  * @param ty_start Starting tile y-coordinate
  * @param ty_end Ending tile y-coordinate
  * @return First matching tile value found, or 0 if none matched, COLLISION_ALL if out of bounds
  */
-UBYTE tile_col_test_range_y(UBYTE tile_mask, UBYTE tx, UBYTE ty_start, UBYTE ty_end) NONBANKED;
+UBYTE tile_col_test_range_y(UBYTE tile_mask, UBYTE tx, UBYTE ty_start, UBYTE ty_end) BANKED;
 
 /**
  * Test for a tile matching mask in a horizontal range from tx_start to tx_end at row ty.
  * Updates globals tile_hit_x and tile_hit_y which can be read afterwards to determine which tile matched
@@ -90,7 +130,7 @@
  * @param tx_start Starting tile x-coordinate
  * @param tx_end Ending tile x-coordinate
  * @return First matching tile value found, or 0 if none matched, COLLISION_ALL if out of bounds
  */
-UBYTE tile_col_test_range_x(UBYTE tile_mask, UBYTE ty, UBYTE tx_start, UBYTE tx_end) NONBANKED;
+UBYTE tile_col_test_range_x(UBYTE tile_mask, UBYTE ty, UBYTE tx_start, UBYTE tx_end) BANKED;
 
 #endif
