Index: src/states/pointnclick.c
===================================================================
--- src/states/pointnclick.c
+++ src/states/pointnclick.c
@@ -10,16 +10,18 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "collision_ex.h"
 
 #ifndef POINT_N_CLICK_CAMERA_DEADZONE
 #define POINT_N_CLICK_CAMERA_DEADZONE 24
 #endif
 
 UBYTE last_hit_trigger = MAX_TRIGGERS;
 
 void pointnclick_init(void) BANKED {
+    COLLISION_EX_INIT();
     camera_offset_x = 0;
     camera_offset_y = 0;
     camera_deadzone_x = POINT_N_CLICK_CAMERA_DEADZONE;
     camera_deadzone_y = POINT_N_CLICK_CAMERA_DEADZONE;
@@ -27,8 +29,9 @@
     actor_set_anim(&PLAYER, ANIM_CURSOR);
 }
 
 void pointnclick_update(void) BANKED {
+    COLLISION_EX_UPDATE();
     UBYTE angle, hit_trigger, is_hover_trigger, is_hover_actor;
     actor_t *hit_actor;
 
     player_moving = FALSE;
