Index: src\core\collision.c
===================================================================
--- src\core\collision.c
+++ src\core\collision.c
@@ -1,7 +1,8 @@
 #pragma bank 255
 
 #include "collision.h"
+#include "scene_transition.h"
 
 UBYTE tile_hit_x = 0;
 UBYTE tile_hit_y = 0;
 
@@ -9,9 +10,9 @@
     tile_hit_x = tx;
     tile_hit_y = ty_start;
 
     if (tile_hit_x >= image_tile_width || tile_hit_y >= image_tile_height) {
-      return COLLISION_ALL & tile_mask ? COLLISION_ALL : 0;
+      return (!scene_transition_enabled && (COLLISION_ALL & tile_mask)) ? COLLISION_ALL : 0;
     }
 
     UBYTE _save = CURRENT_BANK;
     UBYTE inc = UBYTE_LESS_THAN(ty_start, ty_end);
@@ -34,9 +35,9 @@
             tile_ptr -= image_tile_width;
         }
         if (tile_hit_y >= image_tile_height) {
           SWITCH_ROM(_save);
-          return COLLISION_ALL & tile_mask ? COLLISION_ALL : 0;
+          return (!scene_transition_enabled && (COLLISION_ALL & tile_mask)) ? COLLISION_ALL : 0;
         }
     }
     SWITCH_ROM(_save);
     return 0;
@@ -46,9 +47,9 @@
     tile_hit_x = tx_start;
     tile_hit_y = ty;
 
     if (tile_hit_x >= image_tile_width || tile_hit_y >= image_tile_height) {
-      return COLLISION_ALL & tile_mask ? COLLISION_ALL : 0;
+      return (!scene_transition_enabled && (COLLISION_ALL & tile_mask)) ? COLLISION_ALL : 0;
     }
 
     UBYTE _save = CURRENT_BANK;
     UBYTE inc = UBYTE_LESS_THAN(tx_start, tx_end);
@@ -71,9 +72,9 @@
             tile_ptr--;
         }
         if (tile_hit_x >= image_tile_width) {
           SWITCH_ROM(_save);
-          return COLLISION_ALL & tile_mask ? COLLISION_ALL : 0;
+          return (!scene_transition_enabled && (COLLISION_ALL & tile_mask)) ? COLLISION_ALL : 0;
         }
     }
     SWITCH_ROM(_save);
     return 0;
