Index: src\core\ui.c
===================================================================
--- src\core\ui.c
+++ src\core\ui.c
@@ -1,5 +1,5 @@
-// must be in the same bank with ui_a.s
+﻿// must be in the same bank with ui_a.s
 #pragma bank 1
 
 #include <string.h>
 
@@ -34,8 +34,9 @@
 UBYTE win_speed;
 
 UBYTE text_drawn;
 UBYTE current_text_speed;
+UBYTE current_text_layer;
 
 UBYTE text_options;
 UBYTE text_in_speed;
 UBYTE text_out_speed;
@@ -106,8 +107,9 @@
     text_draw_speed             = 1;
     current_text_speed          = 0;
 
     ui_dest_ptr = ui_dest_base  = (text_render_base_addr = GetWinAddr()) + 32 + 1;
+    current_text_layer          = TEXT_LAYER_WIN;
 
     text_scroll_addr            = GetWinAddr();
     text_scroll_width           = 20;
     text_scroll_height          = 8;
@@ -222,9 +224,8 @@
 void ui_print_shift_char(void * dest, const void * src, UBYTE bank) OLDCALL;
 UWORD ui_print_make_mask_lr(UBYTE width, UBYTE ofs) OLDCALL;
 UWORD ui_print_make_mask_rl(UBYTE width, UBYTE ofs) OLDCALL;
 void ui_swap_tiles(void);
-
 UBYTE ui_print_render(const unsigned char ch) {
     UBYTE letter = (vwf_current_font_desc.attr & FONT_RECODE) ? ReadBankedUBYTE(vwf_current_font_desc.recode_table + (ch & vwf_current_font_desc.mask), vwf_current_font_bank) : ch;
     const UBYTE * bitmap = vwf_current_font_desc.bitmaps + letter * 16u;
     if (vwf_current_font_desc.attr & FONT_VWF) {
@@ -303,9 +304,13 @@
         // current char pointer
         ui_text_ptr = ui_text_data;
         // VRAM destination
         if ((text_options & TEXT_OPT_PRESERVE_POS) == 0) {
-            ui_dest_base = text_render_base_addr + 32 + 1;                  // gotoxy(1,1)
+            if (current_text_layer == TEXT_LAYER_BKG){
+                ui_dest_base = text_render_base_addr + ((((UWORD)bkg_offset_y << 5) + 32) & 1023) + (((UWORD)bkg_offset_x + 1) & 31);
+            } else {
+                ui_dest_base = text_render_base_addr + 32 + 1;                  // gotoxy(1,1)
+            }
             if (vwf_direction == UI_PRINT_RIGHTTOLEFT) ui_dest_base += 17;  // right_to_left initial pos correction
             // initialize current pointer with corrected base value
             ui_dest_ptr = ui_dest_base;
             // tileno destination
@@ -346,9 +351,13 @@
                 break;
             }
             case 0x03:
                 // gotoxy
-                ui_dest_ptr = ui_dest_base = text_render_base_addr + (*++ui_text_ptr - 1u) + (*++ui_text_ptr - 1u) * 32u;
+                if (current_text_layer == TEXT_LAYER_BKG){
+                    ui_dest_ptr = ui_dest_base = text_render_base_addr + ((bkg_offset_x + (UWORD)(*++ui_text_ptr - 1u)) & 31) + (((bkg_offset_y + (UWORD)(*++ui_text_ptr - 1u)) << 5) & 1023);
+                } else {
+                    ui_dest_ptr = ui_dest_base = text_render_base_addr + (*++ui_text_ptr - 1u) + (*++ui_text_ptr - 1u) * 32u;
+                }
                 if (vwf_current_offset) ui_print_reset();
                 break;
             case 0x04: {
                 // relative gotoxy
@@ -424,8 +433,14 @@
                 // escape symbol
                 ui_text_ptr++;
                 // fall down to default
             default:
+                //warp around of vram instead of next line on background
+                if (current_text_layer == TEXT_LAYER_BKG) {
+                    if (((UBYTE)ui_dest_ptr >> 5) != ((UBYTE)ui_dest_base >> 5)){
+                        ui_dest_ptr -= 32u;
+                    }
+                }
                 if (ui_print_render(*ui_text_ptr)) {
                     ui_set_tile(ui_dest_ptr, ui_prev_tile, ui_prev_tile_bank);
                     if (vwf_direction == UI_PRINT_LEFTTORIGHT)  ui_dest_ptr++; else ui_dest_ptr--;
                 }
