Index: src\core\vm_ui.c
===================================================================
--- src\core\vm_ui.c
+++ src\core\vm_ui.c
@@ -1,5 +1,5 @@
-#pragma bank 255
+﻿#pragma bank 255
 
 #include <string.h>
 #include <stdlib.h>
 
@@ -140,8 +140,9 @@
 
 // switch text rendering to window or background
 void vm_switch_text_layer(SCRIPT_CTX * THIS, UBYTE target) OLDCALL BANKED {
     THIS;
+    current_text_layer = target;
     if (target) text_render_base_addr = GetWinAddr(); else text_render_base_addr = GetBkgAddr();
 }
 
 // set position of overlayed window
