Index: src\core\interrupts.c
===================================================================
--- src\core\interrupts.c
+++ src\core\interrupts.c
@@ -24,10 +24,10 @@
 }
 
 void simple_LCD_isr(void) NONBANKED {
     if (LYC_REG == LYC_SYNC_VALUE) {
-        SCX_REG = draw_scroll_x;
-        SCY_REG = draw_scroll_y;
+        SCX_REG = bkg_scroll_x;
+        SCY_REG = bkg_scroll_y;
         if (WY_REG) {
             if (WY_REG < MENU_CLOSED_Y) LYC_REG = WY_REG - 1;
         } else {
             if ((WX_REG == DEVICE_WINDOW_PX_OFFSET_X) && (show_actors_on_overlay == FALSE)) HIDE_SPRITES;
@@ -52,10 +52,10 @@
 
 void fullscreen_LCD_isr(void) NONBANKED {
     if (LYC_REG == LYC_SYNC_VALUE) {
         LCDC_REG &= ~LCDCF_BG8000;
-        SCX_REG = draw_scroll_x;
-        SCY_REG = draw_scroll_y;
+        SCX_REG = bkg_scroll_x;
+        SCY_REG = bkg_scroll_y;
         LYC_REG = (9 * 8) - 1;
     } else {
         while (STAT_REG & STATF_BUSY) ;
         LCDC_REG |= LCDCF_BG8000;
