Index: src\states\adventure.c
===================================================================
--- src\states\adventure.c
+++ src\states\adventure.c
@@ -14,8 +14,9 @@
 #include "data_manager.h"
 #include "events.h"
 #include "rand.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "math.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
@@ -330,8 +331,9 @@
     adv_camera_transitioning = FALSE;
     adv_dash_cooldown_timer = 0;
     adv_knockback_timer = 0;
     adv_push_timer = 0;
+    dynamic_actor_init();
 }
 
 void adventure_update(void) BANKED {
     // State transitions
@@ -483,8 +485,9 @@
             camera_deadzone_y == adv_camera_deadzone_y) {
             adv_camera_transitioning = FALSE;
         }
     }
+    dynamic_actor_update();
 }
 
 static void handle_dir_input(void) {
     WORD target_x = 0;
