Index: src\states\topdown.c
===================================================================
--- src\states\topdown.c
+++ src\states\topdown.c
@@ -11,8 +11,9 @@
 #include "input.h"
 #include "trigger.h"
 #include "math.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 
 #ifndef INPUT_TOPDOWN_INTERACT
 #define INPUT_TOPDOWN_INTERACT INPUT_A
 #endif
@@ -32,16 +33,19 @@
     } else {
         PLAYER.pos.x = SUBPX_SNAP_TILE(PLAYER.pos.x);
         PLAYER.pos.y = SUBPX_SNAP_TILE(PLAYER.pos.y);
     }
+    dynamic_actor_init();
 }
 
 void topdown_update(void) BANKED {
     actor_t *hit_actor;
     UBYTE tile_start, tile_end;
     direction_e new_dir = DIR_NONE;
     static UWORD max_pos = 0;
 
+    dynamic_actor_update();
+
     // Is player on an 8x8px tile?
     if ((topdown_grid == 16 && ON_16PX_GRID(PLAYER.pos)) ||
         (topdown_grid == 8 && ON_8PX_GRID(PLAYER.pos))) {
         // Player landed on an tile
