Index: src\states\adventure.c
===================================================================
--- src\states\adventure.c
+++ src\states\adventure.c
@@ -657,9 +657,8 @@
                         }
                     }
                 }
             }
-            new_x = MIN(image_width_subpx - EXCLUSIVE_OFFSET(PLAYER.bounds.right), new_x);
         } else if (delta.x < 0) {
             UBYTE tile_x = SUBPX_TO_TILE(new_x + PLAYER.bounds.left);
             UBYTE tile = tile_col_test_range_y(COLLISION_RIGHT, tile_x, tile_start, tile_end);
             if (tile) {
