Index: include/gbs_types.h
===================================================================
--- include/gbs_types.h
+++ include/gbs_types.h
@@ -10,8 +10,9 @@
 #include "data/scene_types.h"
 #include "bankdata.h"
 #include "parallax.h"
 #include "collision.h"
+#include "data/states_defines.h"
 
 #define COLLISION_GROUP_NONE 0x0
 #define COLLISION_GROUP_PLAYER 0x1
 #define COLLISION_GROUP_1 0x2
@@ -59,8 +60,31 @@
 
     // Collisions
     uint8_t collision_group;
 
+    //Dynamic actor
+    uint8_t actor_index;
+    uint8_t actor_behavior_id : 4;
+    uint8_t actor_state       : 4;
+    int8_t actor_vel_x;
+    int8_t actor_vel_y;
+#ifdef DYNAMIC_ACTOR_ENABLE_PARENT
+    struct actor_t *actor_parent;
+#if !defined(DYNAMIC_ACTOR_PARENT_MODE) || (DYNAMIC_ACTOR_PARENT_MODE == 2)
+    upoint16_t prev_pos;
+#endif
+#endif
+
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+    uint16_t pos_z;
+    int8_t actor_vel_z;
+#ifdef DYNAMIC_ACTOR_ENABLE_PARENT
+#if !defined(DYNAMIC_ACTOR_PARENT_MODE) || (DYNAMIC_ACTOR_PARENT_MODE == 2)
+    uint16_t prev_pos_z;
+#endif
+#endif
+#endif
+
     // Linked list
     struct actor_t *next;
     struct actor_t *prev;
 } actor_t;
@@ -197,4 +221,6 @@
     palette_entry_t cgb_palette[];
 } palette_t;
 
 #endif
+
+
