Index: src\core\actor.c
===================================================================
--- src\core\actor.c
+++ src\core\actor.c
@@ -16,9 +16,9 @@
 #include "collision.h"
 #include "ui.h"
 #include "vm.h"
 #include "macro.h"
-#include "scene_transition.h"
+#include "continuous_scene.h"
 
 #ifdef STRICT
     #include <gb/bgb_emu.h>
     #include <gb/crash_handler.h>
@@ -71,9 +71,8 @@
     player_moving           = FALSE;
     player_iframes          = 0;
     player_collision_actor  = NULL;
     emote_actor             = NULL;
-
     memset(actors, 0, sizeof(actors));
 }
 
 void player_init(void) BANKED {
@@ -550,6 +549,5 @@
     } else if (player_iframes != 0) {
         player_iframes--;
     }
     player_collision_actor = NULL;
-    check_transition_to_scene_collision();
 }
\ No newline at end of file
