Index: src\core\actor.c
===================================================================
--- src\core\actor.c
+++ src\core\actor.c
@@ -16,8 +16,9 @@
 #include "collision.h"
 #include "ui.h"
 #include "vm.h"
 #include "macro.h"
+#include "data/states_defines.h"
 
 #ifdef STRICT
     #include <gb/bgb_emu.h>
     #include <gb/crash_handler.h>
@@ -201,12 +202,20 @@
     // Render player
     if (!CHK_FLAG(PLAYER.flags, ACTOR_FLAG_HIDDEN | ACTOR_FLAG_DISABLED)) {
         if (CHK_FLAG(PLAYER.flags, ACTOR_FLAG_PINNED)) {
             screen_x = SUBPX_TO_PX(PLAYER.pos.x);
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+            screen_y = SUBPX_TO_PX(PLAYER.pos.y - PLAYER.pos_z);
+#else
             screen_y = SUBPX_TO_PX(PLAYER.pos.y);
+#endif
         } else {
             screen_x = SUBPX_TO_PX(PLAYER.pos.x) - draw_scroll_x;
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+            screen_y = SUBPX_TO_PX(PLAYER.pos.y - PLAYER.pos_z) - draw_scroll_y;
+#else
             screen_y = SUBPX_TO_PX(PLAYER.pos.y) - draw_scroll_y;
+#endif
         }
 
 
         if (!(window_hide_actors && (screen_x + 8 > WX_REG) && (screen_y - 8 > WY_REG))) {
@@ -230,12 +239,20 @@
         }
 
         if (CHK_FLAG(actor->flags, ACTOR_FLAG_PINNED)) {
             screen_x = SUBPX_TO_PX(actor->pos.x);
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+            screen_y = SUBPX_TO_PX(actor->pos.y - actor->pos_z);
+#else
             screen_y = SUBPX_TO_PX(actor->pos.y);
+#endif
         } else {
             screen_x = SUBPX_TO_PX(actor->pos.x) - draw_scroll_x;
+#ifdef DYNAMIC_ACTOR_ENABLE_MOVE_Z
+            screen_y = SUBPX_TO_PX(actor->pos.y - actor->pos_z) - draw_scroll_y;
+#else
             screen_y = SUBPX_TO_PX(actor->pos.y) - draw_scroll_y;
+#endif
         }
 
         if (((window_hide_actors) && (((screen_x + 8) > WX_REG) && ((screen_y - 8) > WY_REG)))) {
             continue;
