Index: src\states\adventure.c
===================================================================
--- src\states\adventure.c
+++ src\states\adventure.c
@@ -15,8 +15,9 @@
 #include "events.h"
 #include "rand.h"
 #include "vm.h"
 #include "math.h"
+#include "meta_tiles.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
 
@@ -656,8 +657,13 @@
                             }
                         }
                     }
                 }
+#ifdef ENABLE_ADV_RIGHT_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_RIGHT);
+            } else {
+                reset_collision_cache(DIR_RIGHT);
+#endif
             }
         } else if (delta.x < 0) {
             UBYTE tile_x = SUBPX_TO_TILE(new_x + PLAYER.bounds.left);
             UBYTE tile = tile_col_test_range_y(COLLISION_RIGHT, tile_x, tile_start, tile_end);
@@ -684,8 +690,13 @@
                             }
                         }
                     }
                 }
+#ifdef ENABLE_ADV_LEFT_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_LEFT);
+            } else {
+                reset_collision_cache(DIR_LEFT);
+#endif
             }
         }
 
     finally_update_x:
@@ -731,8 +742,13 @@
                             }
                         }
                     }
                 }
+#ifdef ENABLE_ADV_DOWN_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_DOWN);
+            } else {
+                reset_collision_cache(DIR_DOWN);
+#endif
             }
         } else if (delta.y < 0) {
             UBYTE tile_y = SUBPX_TO_TILE(new_y + PLAYER.bounds.top);
             UBYTE tile = tile_col_test_range_x(COLLISION_BOTTOM, tile_y, tile_start, tile_end);
@@ -759,8 +775,13 @@
                             }
                         }
                     }
                 }
+#ifdef ENABLE_ADV_UP_COLLISION_METATILE
+                on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_UP);
+            } else {
+                reset_collision_cache(DIR_UP);
+#endif
             }
         }
 
     finally_update_y:
@@ -829,8 +850,11 @@
 
     if (mask & COL_CHECK_TRIGGERS)
     {
         trigger_activate_at_intersection(&PLAYER.bounds, &PLAYER.pos, FALSE);
+#ifdef ENABLE_ADV_ENTER_METATILE
+        metatile_overlap_at_intersection(&PLAYER.bounds, &PLAYER.pos);
+#endif
     }
 }
 
 
