Index: src\core\data_manager.c
===================================================================
--- src\core\data_manager.c
+++ src\core\data_manager.c
@@ -167,37 +167,14 @@
 }
 
 UBYTE do_load_palette(palette_entry_t * dest, const palette_t * palette, UBYTE bank) BANKED {
     UBYTE mask = ReadBankedUBYTE(&palette->mask, bank);
-    palette_entry_t * sour = palette->cgb_palette;
-    for (UBYTE i = mask; (i); i >>= 1, dest++) {
-        if ((i & 1) == 0) continue;
-        MemcpyBanked(dest, sour, sizeof(palette_entry_t), bank);
-        sour++;
-    }
+    dest;
+    palette;
+    bank;
     return mask;
 }
 
-inline void load_bkg_palette(const palette_t * palette, UBYTE bank) {
-    UBYTE mask = do_load_palette(BkgPalette, palette, bank);
-    DMG_palette[0] = ReadBankedUBYTE(palette->palette, bank);
-#ifdef SGB
-    if (_is_SGB) {
-        UBYTE sgb_palettes = SGB_PALETTES_NONE;
-        if (mask & 0b00110000) sgb_palettes |= SGB_PALETTES_01;
-        if (mask & 0b11000000) sgb_palettes |= SGB_PALETTES_23;
-        SGBTransferPalettes(sgb_palettes);
-    }
-#endif
-}
-
-inline void load_sprite_palette(const palette_t * palette, UBYTE bank) {
-    do_load_palette(SprPalette, palette, bank);
-    UWORD data = ReadBankedUWORD(palette->palette, bank);
-    DMG_palette[1] = (UBYTE)data;
-    DMG_palette[2] = (UBYTE)(data >> 8);
-}
-
 UBYTE load_scene(const scene_t * scene, UBYTE bank, UBYTE init_data) BANKED {
     UBYTE i;
     scene_t scn;
 
@@ -223,11 +200,8 @@
 
     // Load background + tiles
     load_background(scn.background.ptr, scn.background.bank);
 
-    load_bkg_palette(scn.palette.ptr, scn.palette.bank);
-    load_sprite_palette(scn.sprite_palette.ptr, scn.sprite_palette.bank);
-
     // Copy parallax settings
     memcpy(&parallax_rows, &scn.parallax_rows, sizeof(parallax_rows));
     if (scn.parallax_rows[0].next_y == 0) {
         scene_LCD_type = (scene_type == SCENE_TYPE_LOGO) ? LCD_fullscreen : LCD_simple;
@@ -247,10 +221,17 @@
             allocated_sprite_tiles = player_sprite_len;
         } else {
     #endif
             PLAYER.sprite = scn.player_sprite;
+            PLAYER.reserve_tiles = scn.reserve_tiles;
+            PLAYER.using_sprite_buffer = 0;
             UBYTE n_loaded = load_sprite(PLAYER.base_tile = 0, scn.player_sprite.ptr, scn.player_sprite.bank);
-            allocated_sprite_tiles = player_sprite_len = (n_loaded > scn.reserve_tiles) ? n_loaded : scn.reserve_tiles;
+            allocated_sprite_tiles = (n_loaded > scn.reserve_tiles) ? n_loaded : scn.reserve_tiles;
+            if (scn.reserve_tiles){
+                //double allocated sprite tiles for buffer
+                allocated_sprite_tiles = allocated_sprite_tiles << 1;
+            }
+            player_sprite_len = allocated_sprite_tiles;
             if (!vm_pop_scene_stack_state) {
                 load_animations(scn.player_sprite.ptr, scn.player_sprite.bank, ANIM_SET_DEFAULT, PLAYER.animations);
                 load_bounds(scn.player_sprite.ptr, scn.player_sprite.bank, &PLAYER.bounds);
             }
@@ -302,10 +283,11 @@
             for (i = actors_len - 1; i != 0; i--, actor++) {
                 if (actor->reserve_tiles) {
                     // exclusive sprites allocated separately to avoid overwriting if modified
                     actor->base_tile = allocated_sprite_tiles;
+                    actor->using_sprite_buffer = 0;
                     UBYTE n_loaded = load_sprite(allocated_sprite_tiles, actor->sprite.ptr, actor->sprite.bank);
-                    allocated_sprite_tiles += (n_loaded > actor->reserve_tiles) ? n_loaded : actor->reserve_tiles;
+                    allocated_sprite_tiles += (((n_loaded > actor->reserve_tiles) ? n_loaded : actor->reserve_tiles) << 1);
                 } else {
                     // resolve and set base_tile for each actor
                     UBYTE idx = IndexOfFarPtr(scn.sprites.ptr, scn.sprites.bank, sprites_len, &actor->sprite);
                     actor->base_tile = (idx < sprites_len) ? scene_sprites_base_tiles[idx] : 0;
@@ -354,11 +336,9 @@
     if (triggers_len != 0) {
         MemcpyBanked(&triggers, scn.triggers.ptr, sizeof(trigger_t) * triggers_len, scn.triggers.bank);
     }
 
-    if (!vm_pop_scene_stack_state) {
-        scene_transition_reset();
-    }
+    scene_transition_reset();
     scroll_reset();
     scroll_offset_x = 0;
     scroll_offset_y = 0;
     trigger_reset();
