Index: src\states\pointnclick.c
===================================================================
--- src\states\pointnclick.c
+++ src\states\pointnclick.c
@@ -10,10 +10,10 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "meta_tiles.h"
-#include "continuous_scene.h"
 
 #ifndef POINT_N_CLICK_CAMERA_DEADZONE
 #define POINT_N_CLICK_CAMERA_DEADZONE 24
 #endif
@@ -26,8 +26,9 @@
     camera_deadzone_x = POINT_N_CLICK_CAMERA_DEADZONE;
     camera_deadzone_y = POINT_N_CLICK_CAMERA_DEADZONE;
     PLAYER.dir = DIR_RIGHT;
     actor_set_anim(&PLAYER, ANIM_CURSOR);
+    dynamic_actor_init();
 }
 
 void pointnclick_update(void) BANKED {
     UBYTE angle, hit_trigger, is_hover_trigger, is_hover_actor;
@@ -79,12 +80,9 @@
         } else if (PLAYER.pos.y + PLAYER.bounds.bottom > image_height_subpx) {
             PLAYER.pos.y = image_height_subpx - EXCLUSIVE_OFFSET(PLAYER.bounds.bottom);
         }
     }
-    //Check scene transition
-    if (check_transition_to_scene_collision()) {
-        return;
-    }
+
     // Check for trigger collisions
     hit_trigger = trigger_at_intersection(&PLAYER.bounds, &PLAYER.pos);
 #ifdef ENABLE_POINTNCLICK_ENTER_METATILE
     metatile_overlap_at_intersection(&PLAYER.bounds, &PLAYER.pos);
@@ -114,5 +112,6 @@
             // Run trigger script
             trigger_interact(hit_trigger);
         }
     }
+    dynamic_actor_update();
 }
