Index: src/states/topdown.c
===================================================================
--- src/states/topdown.c
+++ src/states/topdown.c
@@ -11,10 +11,10 @@
 #include "input.h"
 #include "trigger.h"
 #include "math.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "meta_tiles.h"
-#include "continuous_scene.h"
 
 #ifndef INPUT_TOPDOWN_INTERACT
 #define INPUT_TOPDOWN_INTERACT INPUT_A
 #endif
@@ -34,11 +34,14 @@
     } else {
         PLAYER.pos.x = SUBPX_SNAP_TILE(PLAYER.pos.x);
         PLAYER.pos.y = SUBPX_SNAP_TILE(PLAYER.pos.y);
     }
+    dynamic_actor_init();
 }
 
 void topdown_update(void) BANKED {
+
+    dynamic_actor_update();
     actor_t *hit_actor;
     UBYTE tile_start, tile_end;
     direction_e new_dir = DIR_NONE;
     static UWORD max_pos = 0;
@@ -49,23 +52,15 @@
         // Player landed on an tile
         // so stop movement for now
         player_moving = FALSE;
 
-        //Check scene transition
-        if (check_transition_to_scene_collision()) {
-            return;
-        }
-
         // Check for trigger collisions
         if (trigger_activate_at_intersection(&PLAYER.bounds, &PLAYER.pos, FALSE)) {
             // Landed on a trigger
             return;
         }
 #ifdef ENABLE_TOPDOWN_ENTER_METATILE
-        if (metatile_overlap_at_intersection(&PLAYER.bounds, &PLAYER.pos)){
-            // Landed on a metatile with an enter event
-            return;
-        }
+        metatile_overlap_at_intersection(&PLAYER.bounds, &PLAYER.pos);
 #endif
         // Check input to set player movement
         if (INPUT_RECENT_LEFT) {
             player_moving = TRUE;
@@ -176,23 +171,26 @@
         point_translate_dir(&PLAYER.pos, PLAYER.dir, PLAYER.move_speed);
 
         // Clamp to grid
         if (PLAYER.dir == DIR_RIGHT) {
-            if (PLAYER.pos.x > max_pos && (PLAYER.pos.x - max_pos) < PLAYER.move_speed) {
+            if (PLAYER.pos.x > max_pos) {
                 PLAYER.pos.x = max_pos;
             }
         } else if (PLAYER.dir == DIR_LEFT) {
-            if (PLAYER.pos.x < max_pos && (max_pos - PLAYER.pos.x) < PLAYER.move_speed) {
+            if (PLAYER.pos.x < max_pos) {
                 PLAYER.pos.x = max_pos;
+            } else if (max_pos == 0 && PLAYER.pos.x >= (UWORD_MAX - PLAYER.move_speed)) {
+                PLAYER.pos.x = 0;
             }
         } else if (PLAYER.dir == DIR_DOWN) {
-            if (PLAYER.pos.y > max_pos && (PLAYER.pos.y - max_pos) < PLAYER.move_speed) {
+            if (PLAYER.pos.y > max_pos) {
                 PLAYER.pos.y = max_pos;
             }
         } else if (PLAYER.dir == DIR_UP) {
-            if (PLAYER.pos.y < max_pos && (max_pos - PLAYER.pos.y) < PLAYER.move_speed) {
+            if (PLAYER.pos.y < max_pos) {
                 PLAYER.pos.y = max_pos;
+            } else if (max_pos == 0 && PLAYER.pos.y >= (UWORD_MAX - PLAYER.move_speed)) {
+                PLAYER.pos.y = 0;
             }
         }
-
     }
 }
