Index: src\core\scroll.c
===================================================================
--- src\core\scroll.c
+++ src\core\scroll.c
@@ -12,8 +12,10 @@
 #include "math.h"
 #include "fade_manager.h"
 #include "parallax.h"
 #include "palette.h"
+#include "meta_tiles.h"
+#include "data/states_defines.h"
 
 // put submap of a large map to screen
 void set_bkg_submap(UINT8 x, UINT8 y, UINT8 w, UINT8 h, const unsigned char *map, UINT8 map_w) OLDCALL;
 
@@ -22,9 +24,8 @@
 void scroll_load_pending_row(void);
 void scroll_load_pending_col(void);
 void scroll_load_row(UBYTE x, UBYTE y);
 void scroll_load_col(UBYTE x, UBYTE y, UBYTE height);
-void scroll_render_rows(INT16 scroll_x, INT16 scroll_y, BYTE row_offset, BYTE n_rows);
 UBYTE scroll_viewport(parallax_row_t * port);
 
 INT16 scroll_x;
 INT16 scroll_y;
@@ -39,11 +40,15 @@
 UBYTE pending_h_x, pending_h_y;
 UBYTE pending_h_i;
 UBYTE pending_w_x, pending_w_y;
 UBYTE pending_w_i;
-INT16 current_row, new_row;
-INT16 current_col, new_col;
+UBYTE current_row, new_row;
+UBYTE current_col, new_col;
 
+UBYTE scroll_render_disabled;
+
+UWORD bkg_address_offset;
+
 static FASTUBYTE _save_bank;
 
 void scroll_init(void) BANKED {
     draw_scroll_x   = 0;
@@ -59,12 +64,10 @@
 
 void scroll_reset(void) BANKED {
     pending_w_i     = 0;
     pending_h_i     = 0;
-    scroll_x        = 0x7FFF;
-    scroll_y        = 0x7FFF;
-
-    game_time       = 0; // was in scroll_render_rows() - that is insane, here is not the best place either
+    scroll_x = 0x400;
+    scroll_y = 0x400;
 }
 
 void scroll_update(void) BANKED {
     INT16 x, y;
@@ -114,13 +117,13 @@
         port->shadow_scx = shift_scroll_x;
         UBYTE shift_col = PX_TO_TILE(shift_scroll_x);
 
         // If column is +/- 1 just render next column
-        if (current_col == new_col - 1) {
+        if (current_col == (UBYTE)(new_col - 1)) {
             // Render right column
             UBYTE x = shift_col - SCREEN_PAD_LEFT + SCREEN_TILE_REFRES_W - 1;
             scroll_load_col(x, port->start_tile, port->tile_height);
-        } else if (current_col == new_col + 1) {
+        } else if (current_col == (UBYTE)(new_col + 1)) {
             // Render left column
             UBYTE x = MAX(0, shift_col - SCREEN_PAD_LEFT);
             scroll_load_col(x, port->start_tile, port->tile_height);
         } else if (current_col != new_col) {
@@ -132,16 +135,16 @@
         // last parallax slice OR no parallax
         port->shadow_scx = draw_scroll_x;
 
         // If column is +/- 1 just render next column
-        if (current_col == new_col - 1) {
+        if (current_col == (UBYTE)(new_col - 1)) {
             // Queue right column
             UBYTE x = new_col - SCREEN_PAD_LEFT + SCREEN_TILE_REFRES_W - 1;
             UBYTE y = MAX(0, MAX((new_row - SCREEN_PAD_TOP), port->start_tile));
             UBYTE full_y = MAX(0, (new_row - SCREEN_PAD_TOP));
             scroll_queue_col(x, y);
             activate_actors_in_col(x, full_y);
-        } else if (current_col == new_col + 1) {
+        } else if (current_col == (UBYTE)(new_col + 1)) {
             // Queue left column
             UBYTE x = MAX(0, new_col - SCREEN_PAD_LEFT);
             UBYTE y = MAX(0, MAX((new_row - SCREEN_PAD_TOP), port->start_tile));
             UBYTE full_y = MAX(0, (new_row - SCREEN_PAD_TOP));
@@ -155,15 +158,15 @@
             scroll_load_pending_col();
         }
 
         // If row is +/- 1 just render next row
-        if (current_row == new_row - 1) {
+        if (current_row == (UBYTE)(new_row - 1)) {
             // Queue bottom row
             UBYTE x = MAX(0, new_col - SCREEN_PAD_LEFT);
             UBYTE y = new_row - SCREEN_PAD_TOP + SCREEN_TILE_REFRES_H - 1;
             scroll_queue_row(x, y);
             activate_actors_in_row(x, y);
-        } else if (current_row == new_row + 1) {
+        } else if (current_row == (UBYTE)(new_row + 1)) {
             // Queue top row
             UBYTE x = MAX(0, new_col - SCREEN_PAD_LEFT);
             UBYTE y = MAX(port->start_tile, new_row - SCREEN_PAD_TOP);
             scroll_queue_row(x, y);
@@ -184,13 +187,15 @@
     scroll_reset();
     scroll_update();
 }
 
-void scroll_render_rows(INT16 scroll_x, INT16 scroll_y, BYTE row_offset, BYTE n_rows) {
+void scroll_render_rows(INT16 scroll_x, INT16 scroll_y, BYTE row_offset, BYTE n_rows) BANKED {
     // Clear pending rows/ columns
     pending_w_i = 0;
     pending_h_i = 0;
 
+    if (scroll_render_disabled) return;
+
     UBYTE x = MAX(0, PX_TO_TILE(scroll_x) - SCREEN_PAD_LEFT);
     UBYTE y = MAX(0, PX_TO_TILE(scroll_y) + row_offset);
 
     for (BYTE i = 0; i != n_rows && y != image_tile_height; ++i, y++) {
@@ -205,16 +210,19 @@
         // render it now before starting next row
         scroll_load_pending_row();
     }
 
+    if (scroll_render_disabled) return;
+
     // Don't queue rows past image height
     if (y >= image_tile_height) {
         return;
     }
 
     pending_w_x = x;
     pending_w_y = y;
     pending_w_i = SCREEN_TILE_REFRES_W;
+    scroll_load_pending_row();
 }
 
 void scroll_queue_col(UBYTE x, UBYTE y) {
     while (pending_h_i) {
@@ -222,82 +230,153 @@
         // render it now before starting next column
         scroll_load_pending_col();
     }
 
+    if (scroll_render_disabled) return;
+
     pending_h_x = x;
     pending_h_y = y;
     pending_h_i = MIN(SCREEN_TILE_REFRES_H, image_tile_height - y);
+    scroll_load_pending_col();
 }
 
-/* Update pending (up to 5) rows */
-void scroll_load_pending_row(void) NONBANKED {
+void load_metatile_row(const UBYTE* from, UBYTE x, UBYTE y, UBYTE width, UBYTE bank) NONBANKED {
     _save_bank = CURRENT_BANK;
-    UBYTE width = MIN(pending_w_i, PENDING_BATCH_SIZE);
+    SWITCH_ROM(bank);
+    UWORD metatile_y_offset = METATILE_Y_OFFSET(y);
+#if METATILE_SIZE == METATILE_SIZE_16
+    UBYTE tile_y_offset = TILE_Y_OFFSET(y);
+#endif
+    width = width + x;
+    for (x; x != width; x++) {
+#if METATILE_SIZE == METATILE_SIZE_16
+        set_vram_byte((UBYTE*)(0x9800 + bkg_address_offset), *(from + get_metatile_tile(metatile_y_offset + METATILE_X_OFFSET(x), tile_y_offset + TILE_X_OFFSET(x))));
+#else
+        set_vram_byte((UBYTE*)(0x9800 + bkg_address_offset), *(from + (UWORD)sram_map_data[(metatile_y_offset + x)]));
+#endif
+        bkg_address_offset = (bkg_address_offset & 0xFFE0) + ((bkg_address_offset + 1) & 31);
+    }
+    SWITCH_ROM(_save_bank);
+}
 
-#ifdef CGB
-    if (_is_CGB) {  // Color Row Load
-        SWITCH_ROM(image_attr_bank);
-        set_bkg_submap_attributes(pending_w_x, pending_w_y, width, 1, image_attr_ptr, image_tile_width);
+void load_metatile_col(const UBYTE* from, UBYTE x, UBYTE y, UBYTE height, UBYTE bank) NONBANKED {
+    _save_bank = CURRENT_BANK;
+    SWITCH_ROM(bank);
+#if METATILE_SIZE == METATILE_SIZE_16
+    UBYTE metatile_x_offset = METATILE_X_OFFSET(x);
+    UBYTE tile_x_offset = TILE_X_OFFSET(x);
+#endif
+    height = height + y;
+    for (y; y != height; y++) {
+#if METATILE_SIZE == METATILE_SIZE_16
+        set_vram_byte((UBYTE*)(0x9800 + bkg_address_offset), *(from + get_metatile_tile(METATILE_Y_OFFSET(y) + metatile_x_offset, TILE_Y_OFFSET(y) + tile_x_offset)));
+#else
+        set_vram_byte((UBYTE*)(0x9800 + bkg_address_offset), *(from + (UWORD)sram_map_data[METATILE_MAP_OFFSET(x, y)]));
+#endif
+        bkg_address_offset = (bkg_address_offset + 32) & 1023;
     }
-#endif
-    // DMG Row Load
-    SWITCH_ROM(image_bank);
-    set_bkg_submap(pending_w_x, pending_w_y, width, 1, image_ptr, image_tile_width);
-
-    pending_w_x += width;
-    pending_w_i -= width;
-
     SWITCH_ROM(_save_bank);
 }
 
-
-void scroll_load_row(UBYTE x, UBYTE y) NONBANKED {
+void set_bkg_submap_banked(const UBYTE* ptr, UBYTE x, UBYTE y, UBYTE width, UBYTE height, UBYTE source_width, UBYTE bank) NONBANKED {
     _save_bank = CURRENT_BANK;
+    SWITCH_ROM(bank);
+    set_bkg_submap(x, y, width, height, ptr, source_width);
+    SWITCH_ROM(_save_bank);
+}
 
+void scroll_load_row(UBYTE x, UBYTE y) {
+    UBYTE width = MIN(SCREEN_TILE_REFRES_W, image_tile_width);
+    // DMG Row Load
+    if (metatile_bank){
+        bkg_address_offset = ((UWORD)get_bkg_xy_addr(x & 31, y & 31)) - 0x9800;
+        load_metatile_row(metatile_ptr, x, y, width, metatile_bank);
+    } else {
+        set_bkg_submap_banked(image_ptr, x, y, width, 1, image_tile_width, image_bank);
+    }
 #ifdef CGB
-    if (_is_CGB) {  // Color Column Load
-        SWITCH_ROM(image_attr_bank);
-        set_bkg_submap_attributes(x, y, SCREEN_TILE_REFRES_W, 1, image_attr_ptr, image_tile_width);
+    if (_is_CGB) {  // Color Row Load
+        VBK_REG = 1;
+        if (metatile_attr_bank){
+            bkg_address_offset = ((UWORD)get_bkg_xy_addr(x & 31, y & 31)) - 0x9800;
+            load_metatile_row(metatile_attr_ptr, x, y, width, metatile_attr_bank);
+        } else {
+            set_bkg_submap_banked(image_attr_ptr, x, y, width, 1, image_tile_width, image_attr_bank);
+        }
+        VBK_REG = 0;
     }
 #endif
+}
+
+/* Update pending (up to 5) rows */
+void scroll_load_pending_row(void) {
+    UBYTE width = MIN(pending_w_i, PENDING_BATCH_SIZE);
     // DMG Row Load
-    SWITCH_ROM(image_bank);
-    set_bkg_submap(x, y, MIN(SCREEN_TILE_REFRES_W, image_tile_width), 1, image_ptr, image_tile_width);
-
-    SWITCH_ROM(_save_bank);
+    if (metatile_bank){
+        bkg_address_offset = ((UWORD)get_bkg_xy_addr(pending_w_x & 31, pending_w_y & 31)) - 0x9800;
+        load_metatile_row(metatile_ptr, pending_w_x, pending_w_y, width, metatile_bank);
+    } else {
+        set_bkg_submap_banked(image_ptr, pending_w_x, pending_w_y, width, 1, image_tile_width, image_bank);
+    }
+#ifdef CGB
+    if (_is_CGB) {  // Color Row Load
+        VBK_REG = 1;
+        if (metatile_attr_bank){
+            bkg_address_offset = ((UWORD)get_bkg_xy_addr(pending_w_x & 31, pending_w_y & 31)) - 0x9800;
+            load_metatile_row(metatile_attr_ptr, pending_w_x, pending_w_y, width, metatile_attr_bank);
+        } else {
+            set_bkg_submap_banked(image_attr_ptr, pending_w_x, pending_w_y, width, 1, image_tile_width, image_attr_bank);
+        }
+        VBK_REG = 0;
+    }
+#endif
+    pending_w_x += width;
+    pending_w_i -= width;
 }
 
-void scroll_load_col(UBYTE x, UBYTE y, UBYTE height) NONBANKED {
-    _save_bank = CURRENT_BANK;
 
+void scroll_load_col(UBYTE x, UBYTE y, UBYTE height) {
+    // DMG Column Load
+    if (metatile_bank){
+        bkg_address_offset = ((UWORD)get_bkg_xy_addr(x & 31, y & 31)) - 0x9800;
+        load_metatile_col(metatile_ptr, x, y, height, metatile_bank);
+    } else {
+        set_bkg_submap_banked(image_ptr, x, y, 1, height, image_tile_width, image_bank);
+    }
 #ifdef CGB
     if (_is_CGB) {  // Color Column Load
-        SWITCH_ROM(image_attr_bank);
-        set_bkg_submap_attributes(x, y, 1, height, image_attr_ptr, image_tile_width);
+        VBK_REG = 1;
+        if (metatile_attr_bank){
+            bkg_address_offset = ((UWORD)get_bkg_xy_addr(x & 31, y & 31)) - 0x9800;
+            load_metatile_col(metatile_attr_ptr, x, y, height, metatile_attr_bank);
+        } else {
+            set_bkg_submap_banked(image_attr_ptr, x, y, 1, height, image_tile_width, image_attr_bank);
+        }
+        VBK_REG = 0;
     }
 #endif
-    // DMG Column Load
-    SWITCH_ROM(image_bank);
-    set_bkg_submap(x, y, 1, height, image_ptr, image_tile_width);
-    SWITCH_ROM(_save_bank);
 }
 
-void scroll_load_pending_col(void) NONBANKED {
-    _save_bank = CURRENT_BANK;
+void scroll_load_pending_col(void) {
     UBYTE height = MIN(pending_h_i, PENDING_BATCH_SIZE);
-
-    SWITCH_ROM(image_bank);
+    // DMG Column Load
+    if (metatile_bank){
+        bkg_address_offset = ((UWORD)get_bkg_xy_addr(pending_h_x & 31, pending_h_y & 31)) - 0x9800;
+        load_metatile_col(metatile_ptr, pending_h_x, pending_h_y, height, metatile_bank);
+    } else {
+        set_bkg_submap_banked(image_ptr, pending_h_x, pending_h_y, 1, height, image_tile_width, image_bank);
+    }
 #ifdef CGB
     if (_is_CGB) {  // Color Column Load
-        SWITCH_ROM(image_attr_bank);
-        set_bkg_submap_attributes(pending_h_x, pending_h_y, 1, height, image_attr_ptr, image_tile_width);
+        VBK_REG = 1;
+        if (metatile_attr_bank){
+            bkg_address_offset = ((UWORD)get_bkg_xy_addr(pending_h_x & 31, pending_h_y & 31)) - 0x9800;
+            load_metatile_col(metatile_attr_ptr, pending_h_x, pending_h_y, height, metatile_attr_bank);
+        } else {
+            set_bkg_submap_banked(image_attr_ptr, pending_h_x, pending_h_y, 1, height, image_tile_width, image_attr_bank);
+        }
+        VBK_REG = 0;
     }
 #endif
-    // DMG Column Load
-    SWITCH_ROM(image_bank);
-    set_bkg_submap(pending_h_x, pending_h_y, 1, height, image_ptr, image_tile_width);
-
     pending_h_y += height;
     pending_h_i -= height;
-
-    SWITCH_ROM(_save_bank);
 }
