Index: src\states\pointnclick.c
===================================================================
--- src\states\pointnclick.c
+++ src\states\pointnclick.c
@@ -10,8 +10,9 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "continuous_scene.h"
 
 #ifndef POINT_N_CLICK_CAMERA_DEADZONE
 #define POINT_N_CLICK_CAMERA_DEADZONE 24
@@ -25,11 +26,14 @@
     camera_deadzone_x = POINT_N_CLICK_CAMERA_DEADZONE;
     camera_deadzone_y = POINT_N_CLICK_CAMERA_DEADZONE;
     PLAYER.dir = DIR_RIGHT;
     actor_set_anim(&PLAYER, ANIM_CURSOR);
+    dynamic_actor_init();
 }
 
 void pointnclick_update(void) BANKED {
+
+    dynamic_actor_update();
     UBYTE angle, hit_trigger, is_hover_trigger, is_hover_actor;
     actor_t *hit_actor;
 
     player_moving = FALSE;
