Index: include\scroll.h
===================================================================
--- include\scroll.h
+++ include\scroll.h
@@ -56,10 +56,11 @@
  */
 void scroll_repaint(void) BANKED;
 
 void scroll_render_rows(INT16 scroll_x, INT16 scroll_y, BYTE row_offset, BYTE n_rows) BANKED;
-
 void scroll_render_cols(INT16 scroll_x, INT16 scroll_y, BYTE col_offset, BYTE n_cols) BANKED;
+void load_metatile_row(const UBYTE* from, UBYTE x, UBYTE y, UBYTE height, UBYTE bank) NONBANKED;
+void load_metatile_col(const UBYTE* from, UBYTE x, UBYTE y, UBYTE height, UBYTE bank) NONBANKED;
 
 /**
  * Get base address of window map
  */
