Index: src\core\data_manager.c
===================================================================
--- src\core\data_manager.c
+++ src\core\data_manager.c
@@ -14,8 +14,9 @@
 #include "ui.h"
 #include "palette.h"
 #include "data/spritesheet_none.h"
 #include "data/data_bootstrap.h"
+#include "meta_tiles.h"
 #include "data/states_defines.h"
 #include "scene_transition.h"
 #include "macro.h"
 
@@ -346,8 +347,9 @@
     if (triggers_len != 0) {
         MemcpyBanked(&triggers, scn.triggers.ptr, sizeof(trigger_t) * triggers_len, scn.triggers.bank);
     }
 
+    metatile_reset();
     scene_transition_reset();
     scroll_reset();
     scroll_offset_x = 0;
     scroll_offset_y = 0;
