Index: src\states\shmup.c
===================================================================
--- src\states\shmup.c
+++ src\states\shmup.c
@@ -12,10 +12,10 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "meta_tiles.h"
-#include "continuous_scene.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
 
@@ -145,12 +145,15 @@
             PX_TO_SUBPX(SCREEN_HEIGHT_HALF),
             image_height_subpx - PX_TO_SUBPX(SCREEN_HEIGHT_HALF)
         );
     });
+    dynamic_actor_init();
 }
 
 void shmup_update(void) BANKED {
 
+    dynamic_actor_update();
+
     actor_t *hit_actor;
     UBYTE tile_start, tile_end;
     direction_e new_dir = DIR_NONE;
     UBYTE angle;
@@ -270,8 +273,10 @@
                 if (tile) {
                     PLAYER.pos.y = TILE_TO_SUBPX(tile_y) - EXCLUSIVE_OFFSET(PLAYER.bounds.bottom);
 #ifdef ENABLE_DOWN_COLLISION_METATILE
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_DOWN);
+                } else {
+                    reset_collision_cache(DIR_DOWN);
 #endif
                 }
             } else { // Moving up
                 UBYTE tile_y = SUBPX_TO_TILE(PLAYER.pos.y + PLAYER.bounds.top);
@@ -279,8 +284,10 @@
                 if (tile) {
                     PLAYER.pos.y = TILE_TO_SUBPX(tile_y + 1) - PLAYER.bounds.top;
 #ifdef ENABLE_UP_COLLISION_METATILE
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_UP);
+                } else {
+                    reset_collision_cache(DIR_UP);
 #endif
                 }
             }
         }
@@ -295,8 +302,10 @@
                 if (tile) {
                     PLAYER.pos.x = TILE_TO_SUBPX(tile_x) - EXCLUSIVE_OFFSET(PLAYER.bounds.right);
 #ifdef ENABLE_RIGHT_COLLISION_METATILE
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_RIGHT);
+                } else {
+                    reset_collision_cache(DIR_RIGHT);
 #endif
                 }
             } else { // Moving left
                 UBYTE tile_x = SUBPX_TO_TILE(PLAYER.pos.x + PLAYER.bounds.left);
@@ -304,8 +313,10 @@
                 if (tile) {
                     PLAYER.pos.x = TILE_TO_SUBPX(tile_x + 1) - PLAYER.bounds.left;
 #ifdef ENABLE_LEFT_COLLISION_METATILE
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_LEFT);
+                } else {
+                    reset_collision_cache(DIR_LEFT);
 #endif
                 }
             }
         }
@@ -388,8 +399,10 @@
                 })
 #ifdef ENABLE_SHMUP_RIGHT_COLLISION_METATILE
                 if (tile) {
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_RIGHT);
+                } else {
+                    reset_collision_cache(DIR_RIGHT);
                 }
 #endif
             } else { // DIR_LEFT
                 UBYTE tile_x = SUBPX_TO_TILE(PLAYER.pos.x + PLAYER.bounds.left);
@@ -401,8 +414,10 @@
                 })
 #ifdef ENABLE_SHMUP_LEFT_COLLISION_METATILE
                 if (tile) {
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_LEFT);
+                } else {
+                    reset_collision_cache(DIR_LEFT);
                 }
 #endif
             }
         } else {
@@ -418,8 +433,10 @@
                 })
 #ifdef ENABLE_SHMUP_DOWN_COLLISION_METATILE
                 if (tile) {
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_DOWN);
+                } else {
+                    reset_collision_cache(DIR_DOWN);
                 }
 #endif
             } else { // DIR_UP
                 UBYTE tile_y = SUBPX_TO_TILE(PLAYER.pos.y + PLAYER.bounds.top);
@@ -431,8 +448,10 @@
                 })
 #ifdef ENABLE_SHMUP_UP_COLLISION_METATILE
                 if (tile) {
                     on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_UP);
+                } else {
+                    reset_collision_cache(DIR_UP);
                 }
 #endif
             }
         }
@@ -481,12 +500,8 @@
     }
 #endif
 
     if (IS_FRAME_ODD) {
-         //Check scene transition
-        if (check_transition_to_scene_collision()) {
-            return;
-        }
         // Check for trigger collisions on odd frames
 #if SHOOTER_TRIGGER_ACTIVATION == ON_PLAYER_COLLISION
         if (trigger_activate_at_intersection(&PLAYER.bounds, &PLAYER.pos, FALSE)) {
             // Landed on a trigger
