Index: src\states\platform.c
===================================================================
--- src\states\platform.c
+++ src\states\platform.c
@@ -14,8 +14,9 @@
 #include "math.h"
 #include "scroll.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "macro.h"
 #include "continuous_scene.h"
 
 // Feature Flags --------------------------------------------------------------
@@ -581,8 +582,9 @@
     if (PLAYER.dir == DIR_UP || PLAYER.dir == DIR_DOWN || PLAYER.dir == DIR_NONE)
     {
         PLAYER.dir = DIR_RIGHT;
     }
+    dynamic_actor_init();
 }
 
 void platform_update(void) BANKED
 {
@@ -809,8 +811,9 @@
     if ((plat_state != DASH_STATE) && (camera_deadzone_x > plat_camera_deadzone_x))
     {
         camera_deadzone_x--;
     }
+    dynamic_actor_update();
 }
 
 static void player_set_jump_anim(void)
 {
