Index: src\states\shmup.c
===================================================================
--- src\states\shmup.c
+++ src\states\shmup.c
@@ -12,8 +12,9 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 #include "continuous_scene.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
@@ -144,12 +145,15 @@
             PX_TO_SUBPX(SCREEN_HEIGHT_HALF),
             image_height_subpx - PX_TO_SUBPX(SCREEN_HEIGHT_HALF)
         );
     });
+    dynamic_actor_init();
 }
 
 void shmup_update(void) BANKED {
 
+    dynamic_actor_update();
+
     actor_t *hit_actor;
     UBYTE tile_start, tile_end;
     direction_e new_dir = DIR_NONE;
     UBYTE angle;
