Index: src\states\adventure.c
===================================================================
--- src\states\adventure.c
+++ src\states\adventure.c
@@ -16,9 +16,8 @@
 #include "rand.h"
 #include "vm.h"
 #include "math.h"
 #include "meta_tiles.h"
-#include "continuous_scene.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
 
@@ -660,8 +659,10 @@
                     }
                 }
 #ifdef ENABLE_ADV_RIGHT_COLLISION_METATILE
                 on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_RIGHT);
+            } else {
+                reset_collision_cache(DIR_RIGHT);
 #endif
             }
         } else if (delta.x < 0) {
             UBYTE tile_x = SUBPX_TO_TILE(new_x + PLAYER.bounds.left);
@@ -691,8 +692,10 @@
                     }
                 }
 #ifdef ENABLE_ADV_LEFT_COLLISION_METATILE
                 on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_LEFT);
+            } else {
+                reset_collision_cache(DIR_LEFT);
 #endif
             }
         }
 
@@ -741,8 +744,10 @@
                     }
                 }
 #ifdef ENABLE_ADV_DOWN_COLLISION_METATILE
                 on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_DOWN);
+            } else {
+                reset_collision_cache(DIR_DOWN);
 #endif
             }
         } else if (delta.y < 0) {
             UBYTE tile_y = SUBPX_TO_TILE(new_y + PLAYER.bounds.top);
@@ -772,8 +777,10 @@
                     }
                 }
 #ifdef ENABLE_ADV_UP_COLLISION_METATILE
                 on_player_metatile_collision(tile_hit_x, tile_hit_y, DIR_UP);
+            } else {
+                reset_collision_cache(DIR_UP);
 #endif
             }
         }
 
@@ -840,13 +847,8 @@
             }
         }
     }
 
-    //Check scene transition
-    if (check_transition_to_scene_collision()) {
-        return;
-    }
-
     if (mask & COL_CHECK_TRIGGERS)
     {
         trigger_activate_at_intersection(&PLAYER.bounds, &PLAYER.pos, FALSE);
 #ifdef ENABLE_ADV_ENTER_METATILE
