Index: include/gbs_types.h
===================================================================
--- include/gbs_types.h
+++ include/gbs_types.h
@@ -10,8 +10,9 @@
 #include "data/scene_types.h"
 #include "bankdata.h"
 #include "parallax.h"
 #include "collision.h"
+#include "data/states_defines.h"
 
 #define COLLISION_GROUP_NONE 0x0
 #define COLLISION_GROUP_PLAYER 0x1
 #define COLLISION_GROUP_1 0x2
@@ -58,8 +59,9 @@
     uint16_t hscript_update, hscript_hit;
 
     // Collisions
     uint8_t collision_group;
+    uint8_t using_sprite_buffer;
 
     // Linked list
     struct actor_t *next;
     struct actor_t *prev;
