Index: src\core\data_manager.c
===================================================================
--- src\core\data_manager.c
+++ src\core\data_manager.c
@@ -218,10 +218,17 @@
             allocated_sprite_tiles = player_sprite_len;
         } else {
     #endif
             PLAYER.sprite = scn.player_sprite;
+            PLAYER.reserve_tiles = scn.reserve_tiles;
+            PLAYER.using_sprite_buffer = 0;
             UBYTE n_loaded = load_sprite(PLAYER.base_tile = 0, scn.player_sprite.ptr, scn.player_sprite.bank);
-            allocated_sprite_tiles = player_sprite_len = (n_loaded > scn.reserve_tiles) ? n_loaded : scn.reserve_tiles;
+            allocated_sprite_tiles = (n_loaded > scn.reserve_tiles) ? n_loaded : scn.reserve_tiles;
+            if (scn.reserve_tiles){
+                //double allocated sprite tiles for buffer
+                allocated_sprite_tiles = allocated_sprite_tiles << 1;
+            }
+            player_sprite_len = allocated_sprite_tiles;
             load_animations(scn.player_sprite.ptr, scn.player_sprite.bank, ANIM_SET_DEFAULT, PLAYER.animations);
             load_bounds(scn.player_sprite.ptr, scn.player_sprite.bank, &PLAYER.bounds);
     #ifdef DISABLE_PLAYER_SPRITE_LOAD_ON_TRANSITION
         }
@@ -271,10 +278,11 @@
             for (i = actors_len - 1; i != 0; i--, actor++) {
                 if (actor->reserve_tiles) {
                     // exclusive sprites allocated separately to avoid overwriting if modified
                     actor->base_tile = allocated_sprite_tiles;
+                    actor->using_sprite_buffer = 0;
                     UBYTE n_loaded = load_sprite(allocated_sprite_tiles, actor->sprite.ptr, actor->sprite.bank);
-                    allocated_sprite_tiles += (n_loaded > actor->reserve_tiles) ? n_loaded : actor->reserve_tiles;
+                    allocated_sprite_tiles += (((n_loaded > actor->reserve_tiles) ? n_loaded : actor->reserve_tiles) << 1);
                 } else {
                     // resolve and set base_tile for each actor
                     UBYTE idx = IndexOfFarPtr(scn.sprites.ptr, scn.sprites.bank, sprites_len, &actor->sprite);
                     actor->base_tile = (idx < sprites_len) ? scene_sprites_base_tiles[idx] : 0;
