Index: include\collision.h
===================================================================
--- include\collision.h
+++ include\collision.h
@@ -20,8 +20,10 @@
 
 extern UBYTE tile_hit_x;
 extern UBYTE tile_hit_y;
 
+extern UBYTE scene_transition_enabled;
+
 /**
  * Check if point is within positioned bounding box.
  *
  * @param bb Pointer to bounding box
@@ -63,9 +65,9 @@
  */
 inline UBYTE tile_at(UBYTE tx, UBYTE ty) {
     if ((tx < image_tile_width) && (ty < image_tile_height))
         return ReadBankedUBYTE(collision_ptr + (ty * (UINT16)image_tile_width) + tx, collision_bank);
-    return COLLISION_ALL;
+    return (scene_transition_enabled)? 0: COLLISION_ALL;
 }
 
 /**
  * Test for a tile matching mask in a vertical range from ty_start to ty_end at column tx.
