Index: include\camera.h
===================================================================
--- include\camera.h
+++ include\camera.h
@@ -16,12 +16,12 @@
 #define CAMERA_LOCK_Y_MIN_FLAG 0x10
 #define CAMERA_LOCK_Y_MAX_FLAG 0x20
 #define CAMERA_LOCK_FLAG (CAMERA_LOCK_X_FLAG | CAMERA_LOCK_Y_FLAG)
 
-extern UINT16 camera_x;
-extern UINT16 camera_y;
-extern UINT16 camera_clamp_x;
-extern UINT16 camera_clamp_y;
+extern INT16 camera_x;
+extern INT16 camera_y;
+extern INT16 camera_clamp_x;
+extern INT16 camera_clamp_y;
 extern BYTE camera_offset_x;
 extern BYTE camera_offset_y;
 extern BYTE camera_deadzone_x;
 extern BYTE camera_deadzone_y;
