Index: src\states\platform.c
===================================================================
--- src\states\platform.c
+++ src\states\platform.c
@@ -15,8 +15,9 @@
 #include "scroll.h"
 #include "trigger.h"
 #include "vm.h"
 #include "macro.h"
+#include "dynamic_actor.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
 
@@ -581,8 +582,10 @@
     if (PLAYER.dir == DIR_UP || PLAYER.dir == DIR_DOWN || PLAYER.dir == DIR_NONE)
     {
         PLAYER.dir = DIR_RIGHT;
     }
+
+    dynamic_actor_init();
 }
 
 void platform_update(void) BANKED
 {
@@ -809,8 +812,10 @@
     if ((plat_state != DASH_STATE) && (camera_deadzone_x > plat_camera_deadzone_x))
     {
         camera_deadzone_x--;
     }
+
+    dynamic_actor_update();
 }
 
 static void player_set_jump_anim(void)
 {
@@ -1311,10 +1316,10 @@
 
             tile = tile_col_test_range_x(COLLISION_TOP, new_tile_y, tile_x_start, tile_x_end);
             if (tile) {
 #ifdef FEAT_PLATFORM_DROP_THROUGH
-                // Only drop if every collision under the player can be dropped through
-                if (plat_drop_timer && !tile_col_test_range_x(COLLISION_BOTTOM, new_tile_y, tile_x_start, tile_x_end))
+                // Only drop through platforms without a bottom collision
+                if (plat_drop_timer && !(tile & COLLISION_BOTTOM))
                 {
                     goto finally_update_y;
                 }
 #endif
