Index: src\states\shmup.c
===================================================================
--- src\states\shmup.c
+++ src\states\shmup.c
@@ -12,8 +12,9 @@
 #include "game_time.h"
 #include "input.h"
 #include "trigger.h"
 #include "vm.h"
+#include "dynamic_actor.h"
 
 // Feature Flags --------------------------------------------------------------
 // Optional feature flags, set in 'state_defines.h'
 
@@ -143,8 +144,9 @@
             PX_TO_SUBPX(SCREEN_HEIGHT_HALF),
             image_height_subpx - PX_TO_SUBPX(SCREEN_HEIGHT_HALF)
         );
     });
+    dynamic_actor_init();
 }
 
 void shmup_update(void) BANKED {
 
@@ -152,8 +154,10 @@
     UBYTE tile_start, tile_end;
     direction_e new_dir = DIR_NONE;
     UBYTE angle;
 
+    dynamic_actor_update();
+
     player_moving = FALSE;
     new_dir = shooter_direction;
     angle = ANGLE_0DEG;
 
